# Attribute Decoder
----
We have to first look through this document before starting we can correctly interpret the columns in our mobile data set. We will be focusing on the 'Mobile' tab in the excel file and apply the neccessary techniques to organise the data. 

## Import(s)

In [1]:
import pandas as pd

## Reading the Excel file

In [2]:
xls = pd.ExcelFile('Attribute Decoder.xlsx')

In [3]:
tags = pd.read_excel(xls, 'Mobile')

## Looking through the Attribute Decoder

In [4]:
# first look at the dataset
tags.head()

\t"Operating System" Value  Unnamed: 2              \t"Features" Value.1  \
0        \t\t"symbian"    2,         NaN   \t\t"expandable memory"      3,   
1        \t\t"windows"    3,         NaN         \t\t"touchscreen"      0,   
2     \t\t"samsung os"    4,         NaN  \t\t"fingerprint sensor"      4,   
3  \t\t"blackberry os"    5,         NaN           \t\t"dustproof"      6,   
4       \t\t"nokia os"    0,         NaN          \t\t"waterproof"      1,   

   Unnamed: 5 \t"Network Connections" Value.2  Unnamed: 8 \t"Memory RAM"  ...  \
0         NaN                \t\t"4g"      0,         NaN      \t\t"4gb"  ...   
1         NaN                \t\t"2g"      1,         NaN      \t\t"2gb"  ...   
2         NaN                \t\t"3g"      2,         NaN    \t\t"1.5gb"  ...   
3         NaN              \t\t"3.5g"       3         NaN     \t\t"16gb"  ...   
4         NaN                     NaN     NaN         NaN    \t\t"512mb"  ...   

  Value.7  Unnamed: 23                       \t"Phone Model" Value.8  \
0      4,          NaN              \t\t"samsung gear sport"   1142,   
1      0,          NaN            \t\t"sony xperia xa ultra"      0,   
2     14,          NaN              \t\t"samsung ua55mu6300"   1998,   
3     15,          NaN  \t\t"apple ipad mini wi fi cellular"   1224,   
4      1,          NaN                 \t\t"coocaa 55s3a12g"   1144,   

   Unnamed: 26      \t"Camera" Value.9  Unnamed: 29  \
0          NaN      \t\t"42mp"      0,          NaN   
1          NaN  \t\t"dua slot"      1,          NaN   
2          NaN      \t\t"5 mp"      2,          NaN   
3          NaN      \t\t"3 mp"      3,          NaN   
4          NaN      \t\t"1 mp"      4,          NaN   

        \t"Phone Screen Size" Value.10  
0       \t\t"4.6 to 5 inches"       0,  
1     \t\t"4.1 to 4.5 inches"       3,  
2  \t\t"less than 3.5 inches"       4,  
3       \t\t"3.6 to 4 inches"       1,  
4  \t\t"more than 5.6 inches"       2,  

[5 rows x 32 columns]

In [5]:
# looking through the columns
tags.columns

Index(['\t"Operating System"', 'Value', 'Unnamed: 2', '\t"Features"',
       'Value.1', 'Unnamed: 5', '\t"Network Connections"', 'Value.2',
       'Unnamed: 8', '\t"Memory RAM"', 'Value.3', 'Unnamed: 11', '\t"Brand"',
       'Value.4', 'Unnamed: 14', '\t"Warranty Period"', 'Value.5',
       'Unnamed: 17', '\t"Storage Capacity"', 'Value.6', 'Unnamed: 20',
       '\t"Color Family"', 'Value.7', 'Unnamed: 23', '\t"Phone Model"',
       'Value.8', 'Unnamed: 26', '\t"Camera"', 'Value.9', 'Unnamed: 29',
       '\t"Phone Screen Size"', 'Value.10'],
      dtype='object')

In [6]:
# checking for nulls 
tags.isnull().sum()

\t"Operating System"       2273
Value                      2273
Unnamed: 2                 2280
\t"Features"               2273
Value.1                    2273
Unnamed: 5                 2280
\t"Network Connections"    2276
Value.2                    2276
Unnamed: 8                 2280
\t"Memory RAM"             2270
Value.3                    2270
Unnamed: 11                2280
\t"Brand"                  2224
Value.4                    2224
Unnamed: 14                2280
\t"Warranty Period"        2266
Value.5                    2266
Unnamed: 17                2280
\t"Storage Capacity"       2262
Value.6                    2262
Unnamed: 20                2280
\t"Color Family"           2254
Value.7                    2254
Unnamed: 23                2280
\t"Phone Model"               0
Value.8                       0
Unnamed: 26                2280
\t"Camera"                 2265
Value.9                    2265
Unnamed: 29                2280
\t"Phone Screen Size"      2274
Value.10

There are many empty rows in the data set except for 'Phone Model'. This is because only 'Phone Model' has 2280 unique rows while the other columns have less unqiue rows.

In [7]:
# extracting the blank columns
unnamed_cols = [col for col in tags.columns if 'Unnamed' in col]
unnamed_cols

['Unnamed: 2',
 'Unnamed: 5',
 'Unnamed: 8',
 'Unnamed: 11',
 'Unnamed: 14',
 'Unnamed: 17',
 'Unnamed: 20',
 'Unnamed: 23',
 'Unnamed: 26',
 'Unnamed: 29']

In [8]:
# dropping the unnamed cols
tags = tags.drop(columns=unnamed_cols)

In [9]:
# verifying the changes
tags.columns

Index(['\t"Operating System"', 'Value', '\t"Features"', 'Value.1',
       '\t"Network Connections"', 'Value.2', '\t"Memory RAM"', 'Value.3',
       '\t"Brand"', 'Value.4', '\t"Warranty Period"', 'Value.5',
       '\t"Storage Capacity"', 'Value.6', '\t"Color Family"', 'Value.7',
       '\t"Phone Model"', 'Value.8', '\t"Camera"', 'Value.9',
       '\t"Phone Screen Size"', 'Value.10'],
      dtype='object')

In [10]:
# defining dict for new column names
tags_dict = {'\t"Operating System"':'os', 'Value':'os_val',
             '\t"Features"':'features', 'Value.1':'features_val',
             '\t"Network Connections"':'network_con', 'Value.2':'network_con_val',
             '\t"Memory RAM"':'ram', 'Value.3':'ram_val',
             '\t"Brand"':'brand', 'Value.4':'brand_val',
             '\t"Warranty Period"':'warranty', 'Value.5':'warranty_val',
             '\t"Storage Capacity"':'storage', 'Value.6':'storage_val',
             '\t"Color Family"':'color', 'Value.7':'color_val',
             '\t"Phone Model"':'model', 'Value.8':'model_val',
             '\t"Camera"':'camera', 'Value.9':'camera_val',
             '\t"Phone Screen Size"':'screen_size', 'Value.10':'screen_size_val'}

In [11]:
# renaming columns
tags = tags.rename(columns=tags_dict)

In [12]:
# function to remove noise from each entry
def clean(x):
    try:
        return x.split('\t\t"')[1].split('"')[0]
    except:
        return x

# function to remove commas    
def remove_comma(x):
    try:
        return x.strip(',')
    except:
        return x

In [13]:
# cleaning the dataset 
for i in tags.columns:
    tags[i] = tags[i].apply(lambda x : remove_comma(clean(x)))

In [14]:
# verifying the changes
tags.head()

os os_val            features features_val network_con  \
0        symbian      2   expandable memory            3          4g   
1        windows      3         touchscreen            0          2g   
2     samsung os      4  fingerprint sensor            4          3g   
3  blackberry os      5           dustproof            6        3.5g   
4       nokia os      0          waterproof            1         NaN   

  network_con_val    ram ram_val     brand brand_val  ... storage storage_val  \
0               0    4gb       5    google         0  ...   256gb          16   
1               1    2gb       6       htc         1  ...   1.5gb           0   
2               2  1.5gb       0     apple         2  ...   128gb           1   
3               3   16gb       9      wiko         3  ...   512mb           2   
4             NaN  512mb       1  polytron         4  ...    64gb           3   

       color color_val                           model model_val    camera  \
0       blue         4              samsung gear sport      1142      42mp   
1       gold         0            sony xperia xa ultra         0  dua slot   
2      brown        14              samsung ua55mu6300      1998      5 mp   
3  navy blue        15  apple ipad mini wi fi cellular      1224      3 mp   
4     yellow         1                 coocaa 55s3a12g      1144      1 mp   

  camera_val           screen_size screen_size_val  
0          0       4.6 to 5 inches               0  
1          1     4.1 to 4.5 inches               3  
2          2  less than 3.5 inches               4  
3          3       3.6 to 4 inches               1  
4          4  more than 5.6 inches               2  

[5 rows x 22 columns]

In [16]:
tags['warranty']

0       7 months
1       4 months
2       6 months
3       3 months
4       10 years
          ...   
2275         NaN
2276         NaN
2277         NaN
2278         NaN
2279         NaN
Name: warranty, Length: 2280, dtype: object

## Export

In [ ]:
tags.to_csv('tags.csv', index=False)